frontend/src/metabase/visualizations/visualizations/SmartScalar/utils.ts

```diff
@@ -21,36 +20,27 @@ export const formatChangeAutoPrecision = (
     fontWeight,
     width,
   }: { fontFamily: string; fontWeight: number; width: number },
-): string => {
-  for (let fractionDigits = 2; fractionDigits >= 1; --fractionDigits) {
-    const formatted = formatChange(change, {
-      maximumFractionDigits: fractionDigits,
-    });
-
-    const formattedWidth = measureText(formatted, {
-      size: "1rem",
-      family: fontFamily,
-      weight: fontWeight,
-    }).width;
-
-    if (formattedWidth <= width) {
-      return formatted;
-    }
-  }
-
-  return formatChange(change, {
-    maximumFractionDigits: 0,
-  });
-};
+): string =>
+  [2, 1]
+    .map(n => formatChange(change, { maximumFractionDigits: n }))
+    .find(
+      formatted =>
+        measureText(formatted, {
+          size: "1rem",
+          family: fontFamily,
+          weight: fontWeight,
+        }).width <= width,
+    ) ?? formatChange(change, { maximumFractionDigits: 0 });
```

Comment:
a functional refactoring for the next guy who has to look at it.  [existing unit tests](https://github.com/metabase/metabase/blob/67a1d7847390308e52d0f1cc83fbd66eeac8724b/frontend/src/metabase/visualizations/visualizations/SmartScalar/utils.unit.spec.ts#L49-L74) verify that it still works.